### ISSUE TITLE
Moq1002 Analyzer should flag cases of missing constructor parameters

### ISSUE THREAD

**Issue by MattKotsenas** [rjmurillo/moq.analyzers#55]
## Scenario

```csharp
public abstract class AbstractClassWithCtor
{
    protected AbstractClassWithCtor(int a)
    {
    }
}

var mock3 = new Mock<AbstractClassWithCtor>();
```

## Expected Result

- Moq1002

## Actual Result

No diagnostic

**Comment by rjmurillo** [comment]
There is an implicit ctor for the type, so the diagnostic does not fire. The issue you're asking for here is while that is technically correct, it's not what was intended. Are you wanting something else, or is that correct?

**Comment by MattKotsenas** [comment]
What I would expect (jotting down notes):

Because there's an explicit constructor there shouldn't be an implicit constructor. Then we should fail to find a matching constructor like we do for other cases of Moq1002.

This is backed up by trying to instantiate this mock; Castle blows up:

```
Unhandled exception. System.ArgumentException: Can not instantiate proxy of class: AbstractClassWithCtor.
Could not find a parameterless constructor. (Parameter 'constructorArguments')
 ---> System.MissingMethodException: Constructor on type 'Castle.Proxies.AbstractClassWithCtorProxy' not found.
```

**Comment by rjmurillo** [comment]
Yep, in the 'abstract' case that makes sense. I'll look at the analyzer to see if the condition can be added (e.g. IsAbstract && !MatchingCtors()). If it's not already there it's certainly in other analyzers already.

**Comment by rjmurillo** [comment]
This is bailing out on L62 before we check what the mocked type has to offer

https://github.com/rjmurillo/moq.analyzers/blob/12e9e315e153212b72594781c99319b18c545a29/src/Moq.Analyzers/ConstructorArgumentsShouldMatchAnalyzer.cs#L58-L62

**Comment by rjmurillo** [comment]
These should also fail but currently do not because of this issue

https://github.com/rjmurillo/moq.analyzers/blob/12e9e315e153212b72594781c99319b18c545a29/tests/Moq.Analyzers.Test/ConstructorArgumentsShouldMatchAnalyzerTests.cs#L11-L13

The test class is defined as

https://github.com/rjmurillo/moq.analyzers/blob/12e9e315e153212b72594781c99319b18c545a29/tests/Moq.Analyzers.Test/ConstructorArgumentsShouldMatchAnalyzerTests.cs#L62-L68

Which generates

```csharp
.class private auto ansi beforefieldinit Foo
    extends [System.Runtime]System.Object
{
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableContextAttribute::.ctor(uint8) = (
        01 00 01 00 00
    )
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableAttribute::.ctor(uint8) = (
        01 00 00 00 00
    )
    // Methods
    .method public hidebysig specialname rtspecialname 
        instance void .ctor (
            string s
        ) cil managed 
    {
        // Method begins at RVA 0x2050
        // Code size 9 (0x9)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void [System.Runtime]System.Object::.ctor()
        IL_0006: nop
        IL_0007: nop
        IL_0008: ret
    } // end of method Foo::.ctor

    .method public hidebysig specialname rtspecialname 
        instance void .ctor (
            bool b,
            int32 i
        ) cil managed 
    {
        // Method begins at RVA 0x2050
        // Code size 9 (0x9)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void [System.Runtime]System.Object::.ctor()
        IL_0006: nop
        IL_0007: nop
        IL_0008: ret
    } // end of method Foo::.ctor

    .method public hidebysig specialname rtspecialname 
        instance void .ctor (
            valuetype [System.Runtime]System.DateTime[] dates
        ) cil managed 
    {
        .param [1]
            .custom instance void [System.Runtime]System.ParamArrayAttribute::.ctor() = (
                01 00 00 00
            )
        // Method begins at RVA 0x2050
        // Code size 9 (0x9)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void [System.Runtime]System.Object::.ctor()
        IL_0006: nop
        IL_0007: nop
        IL_0008: ret
    } // end of method Foo::.ctor

    .method public hidebysig specialname rtspecialname 
        instance void .ctor (
            class [System.Collections]System.Collections.Generic.List`1<string> l,
            [opt] string s
        ) cil managed 
    {
        .param [2] = "A"
        // Method begins at RVA 0x2050
        // Code size 9 (0x9)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void [System.Runtime]System.Object::.ctor()
        IL_0006: nop
        IL_0007: nop
        IL_0008: ret
    } // end of method Foo::.ctor

} // end of class Foo
```
The generated class does not have a parameterless ctor. Whereas 

```csharp
internal class Bar
{
}
```

Generates

```csharp
.class private auto ansi beforefieldinit Bar
    extends [System.Runtime]System.Object
{
    // Methods
    .method public hidebysig specialname rtspecialname 
        instance void .ctor () cil managed 
    {
        // Method begins at RVA 0x205a
        // Code size 8 (0x8)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void [System.Runtime]System.Object::.ctor()
        IL_0006: nop
        IL_0007: ret
    } // end of method Bar::.ctor

} // end of class Bar
```

Which has a default parameterless ctor
